fff

```diff
@@ -368,6 +369,57 @@ trash() {
     fi
 }
 
+bulk_rename() {
+  local tmp_file_template="${XDG_CACHE_HOME:=${HOME}/.cache}/fff/bulkrename.XXXXX"
+
+  local tmp_file
+  printf -v tmp_file '%q' "$(mktemp "$tmp_file_template")"
+
+  # Write all input file names to temporary file and open it in an editor.
+  printf "%s\n" "${@:1:$#-1}" > "$tmp_file"
+  "${EDITOR:-vi}" "$tmp_file"
+
+  local commands_file
+  printf -v commands_file '%q' "$(mktemp "$tmp_file_template")"
+
+  # Add instructions to the file with commands.
+  printf '%s\n%s\n' \
+    "# This file will be executed when you close the editor." \
+    "# Please double-check everything, clear the file to abort." > "$commands_file"
+
+  local counter=0
+  local changed=0
+  while read -r line || [ -n "$line" ] ; do
+    ((counter++))
+
+    # Trim read line.
+    line="${line#"${line%%[![:space:]]*}"}"
+    line="${line%"${line##*[![:space:]]}"}"
+
+    [ -z "$line" ] && continue
+
+    # Escape source and target arguments of mv command.
+    local source
+    printf -v source '%q' "${@:counter:1}"
+
+    local target
+    printf -v target '%q' "$line"
+
+    [[ $source != "$target" ]] && {
+      printf "%s\n" "mv -vi -- $source $target" >> "$commands_file"
+      changed=1
+    }
+  done < "$tmp_file"
+
+  # Old and new locations are the same for all files, do nothing.
+  ((changed == 0)) && return
+
+  "${EDITOR:-vi}" "$commands_file"
```

Comment:
Add `setup_terminal` below this line to re-hide the cursor and fix other settings after `$EDITOR` exits.